clone/cmd/sumdbclone/internal/client/sumdb.go

```diff
@@ -0,0 +1,164 @@
+func NewSumDB(height int, vkey, url string, c *http.Client) *SumDBClient {
+	return &SumDBClient{
+		height: height,
+		vkey:   vkey,
+		fetcher: &HTTPFetcher{
+			c:       c,
+			baseURL: url,
+		},
+	}
+}
+
+// LatestCheckpoint gets the freshest Checkpoint.
+func (c *SumDBClient) LatestCheckpoint() (*Checkpoint, error) {
+	checkpoint, err := c.fetcher.GetData("/latest")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get /latest Checkpoint; %w", err)
+	}
+
+	verifier, err := note.NewVerifier(c.vkey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create verifier: %w", err)
+	}
+	verifiers := note.VerifierList(verifier)
+
+	note, err := note.Open(checkpoint, verifiers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open note: %w", err)
+	}
+	tree, err := tlog.ParseTree([]byte(note.Text))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse tree: %w", err)
+	}
+
+	return &Checkpoint{Tree: &tree, Raw: checkpoint}, nil
+}
+
+// FullLeavesAtOffset gets the Nth chunk of 2**height leaves.
+func (c *SumDBClient) FullLeavesAtOffset(offset int) ([][]byte, error) {
+	data, err := c.fetcher.GetData(fmt.Sprintf("/tile/%d/data/%s", c.height, c.tilePath(offset)))
+	if err != nil {
+		return nil, err
+	}
+	return dataToLeaves(data), nil
+}
+
+// PartialLeavesAtOffset gets the final tile of incomplete leaves.
+func (c *SumDBClient) PartialLeavesAtOffset(offset, count int) ([][]byte, error) {
+	data, err := c.fetcher.GetData(fmt.Sprintf("/tile/%d/data/%s.p/%d", c.height, c.tilePath(offset), count))
+	if err != nil {
+		return nil, err
+	}
+	return dataToLeaves(data), nil
+}
+
+// tilePath constructs the component of the path which refers to a tile at a
+// given offset. This was copied from the core implementation:
+// https://github.com/golang/mod/blob/346a37af5599be02f125bd8bc0a5e1d33db21ddc/sumdb/tlog/tile.go#L171
+func (c *SumDBClient) tilePath(offset int) string {
+	nStr := fmt.Sprintf("%03d", offset%pathBase)
+	for offset >= pathBase {
+		offset /= pathBase
+		nStr = fmt.Sprintf("x%03d/%s", offset%pathBase, nStr)
+	}
+	return nStr
+}
+
+func dataToLeaves(data []byte) [][]byte {
+	result := make([][]byte, 0)
+	start := 0
+	for i, b := range data {
+		if b == '\n' {
+			if i > start && data[i-1] == '\n' {
+				result = append(result, data[start:i])
+				start = i + 1
+			}
+		}
+	}
+	result = append(result, data[start:])
+	return result
+}
+
+// HTTPFetcher gets the data over HTTP(S).
+type HTTPFetcher struct {
+	c       *http.Client
+	baseURL string
+}
+
+// GetData gets the data.
+func (f *HTTPFetcher) GetData(path string) ([]byte, error) {
+	target := f.baseURL + path
+	resp, err := f.c.Get(target)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			glog.Errorf("resp.Body.Close(): %v", err)
+		}
+	}()
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("GET %v: %v", target, resp.Status)
+	}
+	data, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
```

Comment:
could this just directly `return io.ReadAll(...)` since you're not decorating the error?

Reply:
Done.